### ISSUE TITLE
Incorrect include path when libpqxx is added with add_subdirectory in cmake

### ISSUE THREAD

**Issue by dpolunin** [jtv/libpqxx#127]
Root `CMakeLists.txt` should probably use `CMAKE_CURRENT_SOURCE_DIR` instead of `CMAKE_SOURCE_DIR`.
Also `cmake/config.cmake` should use `${libpqxx_SOURCE_DIR}` and `${libpqxx_BINARY_DIR}`.

**Comment by llarose** [comment]
Hi,

CMAKE_SOURCE_DIR is relative to the root of the source tree.
CMAKE_CURRENT_SOURCE_DIR is the current directory where a (node/leaf)
CMakeLists.txt lives.

I can't talk about  ${libpqxx_SOURCE_DIR}/${libpqxx_BINARY_DIR}, but CMake
has the concept of "out of source build".
When generating make file in a build folder, CMake will use
CMAKE_BINARY_DIR and CMAKE_CURRENT_BINARY_DIR to specify the target folders
in the build folder tree. This is where you'll find all .o and .so, .a,
.dll and executable (cmake targets).

Hope it answer your question.

By the way, I did include pqxx in a complex CMake project. Since by the
time of my configuration, pqxx isn't configured with CMake. I imported pqxx
as a external project and kept the configuration and compilation like it
was already suggested.

Here is a snapshot of my CMakeList code :

execute_process(COMMAND tar -xf
${CMAKE_SOURCE_DIR}/libs/libpqxx-5.0.1.tar.gz -C ${CMAKE_BINARY_DIR}/libs/)
include (ExternalProject)
ExternalProject_Add (
  pqxx
  SOURCE_DIR ${CMAKE_BINARY_DIR}/libs/libpqxx-5.0.1
  CONFIGURE_COMMAND ${CMAKE_BINARY_DIR}/libs/libpqxx-5.0.1/configure
--with-pic --disable-documentation --enable-static --enable-shared -q
  BUILD_IN_SOURCE 1
  BUILD_BYPRODUCTS
${CMAKE_BINARY_DIR}/libs/libpqxx-5.0.1/src/.libs/libpqxx.so
${CMAKE_BINARY_DIR}/libs/libpqxx-5.0.1/src/.libs/libpqxx-5.0.so
  BUILD_COMMAND "make"
  INSTALL_COMMAND ""
)
ExternalProject_Get_Property (pqxx binary_dir source_dir)
set(pqxx_binary_dir ${binary_dir})
set(pqxx_source_dir ${source_dir})
set(pqxx_include_dir ${pqxx_binary_dir}/include)
set(pqxx_link_dir ${pqxx_binary_dir}/src/.libs)

add_library(libpqxx INTERFACE IMPORTED GLOBAL)
set_property(TARGET libpqxx PROPERTY INTERFACE_INCLUDE_DIRECTORIES
${pqxx_include_dir})
set_property(TARGET libpqxx PROPERTY INTERFACE_LINK_LIBRARIES
${pqxx_link_dir}/libpqxx.so)


Then when ever you need pqxx linked to another target, you call :
            add_dependencies( MyTarget pqxx)
      target_link_libraries(MyTarget libpqxx)

Hope it helps.
--
LAL

**Comment by JadeMatrix** [comment]
Is this the segment in question?

```cmake
include_directories(
  ${CMAKE_SOURCE_DIR}/include
  ${PostgreSQL_INCLUDE_DIRS}
)
```

In this case, probably neither `CMAKE_SOURCE_DIR` nor `CMAKE_CURRENT_SOURCE_DIR` should be used, as that makes the path absolute rather than relative (just `include/` is fine).  But really it doesn't hurt, since libpqxx's CMake config isn't exported.  @jtv I was thinking of possibly tackling that next after the `tablereader`/`tablewriter` stuff.

**Comment by dpolunin** [comment]
> Is this the segment in question?

There are two segments in question in CMakeLists.txt. The one you mentioned and this one:
```cmake
set(CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake)
```
In my case `${CMAKE_SOURCE_DIR}` doesn't point to libpqxx root. That's why `include(config)` fails.

**Comment by dpolunin** [comment]
> What's the difference between `${libpqxx_SOURCE_DIR}`/`${libpqxx_BINARY_DIR}` and what `config.cmake` does now?

According to [documentation](https://cmake.org/cmake/help/latest/variable/PROJECT-NAME_SOURCE_DIR.html) `${libpqxx_SOURCE_DIR}` would point to libpqxx source root directory (and `${libpqxx_BINARY_DIR}` to libpqxx build root). `${CMAKE_SOURCE_DIR}` would point to root of my project which at some point uses `add_subdirectory` to include libpqxx. That's why `cmake.config` uses wrong location for `config.h.in` and puts `config.h` in build root instead of libpqxx subdirectory in build.

**Comment by jtv** [comment]
Any chance you could put this in a well-tested pull request?  If I make the changes myself the odds are that I'll mess it up.  Plus, it's a chance to get your name on a commit.  :-)

**Comment by jtv** [comment]
Would love to see a patch!  If you feel impelled to do it, comment things as if for a child though.  :-)

**Comment by abrownsword** [comment]
That would also require me figuring out how to do that, and I’m a relative newb on github and git so there’s a learning curve I haven’t climbed yet.

**Comment by jtv** [comment]
That's me with CMake, basically.  :-)

**Comment by abrownsword** [comment]
I spent 5 minutes trying to figure out how to do a PR, and it just tells me I don't have authorization.  Followed the GitHub help on PRs.  Don't have time to mess around with it much more.  Plus the amount of testing and commenting I can do is minimal -- the modified CMakeLists.txt files (its just the 2 files) work for me as a sub-project, however the originals don't work on my machines because find_package is broken for some reason (I'm not a cmake expert either, and I figured out a nice way around having to figure out find_package, so...).

I'll attach the two files here and someone more git and cmake savvy can "Do The Right Thing":

[CMakeLists.txt](https://github.com/jtv/libpqxx/files/2430829/CMakeLists.txt)
[src/CMakeLists.txt](https://github.com/jtv/libpqxx/files/2430830/CMakeLists.txt)

My commit comment:

Tweak cmake files to make useable as sub-project within another cmake project:
* Switched to using PROJECT_*_DIR instead of CMAKE_*_DIR
* Created SKIP_BUILD_TEST, SKIP_PQXX_SHARED and SKIP_PQXX_STATIC to control what gets build via cmake variables
* Skip doing find_package(PostgreSQL) if its variables are already set
